### ISSUE TITLE
`--isolatedModules` should warn about re-export of type

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#15231]
**TypeScript Version:**  nightly (Version 2.3.0-dev.20170417)

**Code**

**a.ts**

```ts
export const A = 0;
export type B = number;
```

**b.ts**

```ts
export { A, B } from "./a";
```

**Expected behavior:**

An error saying that the emit of `b.ts` depends on the contents of `a.ts`.

**Actual behavior:**

`b.js` is `export { A } from "./a";`.

**Comment by Kovensky** [comment]
Hm... what exactly is `--isolatedModules` supposed to do?

I have been using it as a way of requiring that only files parsed as "module" would be considered valid (i.e. requires at least one "import" or "export" in typescript), though preventing reexporting of types would break things on my project.

I can turn of `--isolatedModules` off if that is the case, but since it seems to do more than what I thought it did, I wonder what would be the side effects.

**Comment by mhegazy** [comment]
It guarantees that the file can be transpiled in isolation. Hence the isolated part. Things  that need cross file information to emit like namespaces spanning multiple files or const enums are not alloewd. This is what you would use if you were using transpileModule API.

**Comment by Kovensky** [comment]
So I assume the absence of `--isolatedModules` is what causes this behavior? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16118#discussion_r113232157

If that is the case, disabling `--isolatedModules` could be problematic in my codebase, as I have multiple _independent_ (non-exported) `interface Props {}` in different files. IIUC, this would cause them to augment each other instead of still being independent.

**Comment by andy-ms** [comment]
`--isolatedModules` won't cause "isolated" type-checking behavior, just isolated transpilation. So multiple globally-declared interfaces will still be merged. Not sure which specific issue you're talking about as the link there just takes me to the issue, not to a particular comment.

**Comment by Kovensky** [comment]
@andy-ms ah, that happens sometimes in github. The thing in question was about `type` / `interface` in `.d.ts` being `export`ed regardless of the presence of `export`, which upon thinking further wouldn't (unintentionally) be further augmentable.
